File: packages/core/src/marginalia_core/application/frontend/gateway.py
"""Frontend gateway protocol."""

from __future__ import annotations

from typing import Protocol

from marginalia_core.application.frontend.capabilities import BackendCapabilities
from marginalia_core.application.frontend.envelopes import FrontendRequest, FrontendResponse
from marginalia_core.application.frontend.events import FrontendEvent


class FrontendGateway(Protocol):
    """Single backend surface for client frontends."""

    def capabilities(self) -> BackendCapabilities: ...

    def execute_command(self, request: FrontendRequest) -> FrontendResponse: ...

    def execute_query(self, request: FrontendRequest) -> FrontendResponse: ...

    def recent_events(self, limit: int = 50) -> tuple[FrontendEvent, ...]: ...

File: crates/marginalia-core/src/frontend/gateway.rs
use crate::frontend::{BackendCapabilities, FrontendEvent, FrontendRequest, FrontendResponse};

pub trait FrontendGateway {
    fn capabilities(&self) -> BackendCapabilities;
    fn execute_command(&mut self, request: FrontendRequest) -> FrontendResponse;
    fn execute_query(&mut self, request: FrontendRequest) -> FrontendResponse;
    fn recent_events(&self, limit: usize) -> Vec<FrontendEvent>;
}
